rnd/autogpt_server/autogpt_server/executor/manager.py

```diff
@@ -228,44 +240,94 @@ def get_agent_server_client() -> "AgentServer":
 
 
 class Executor:
-    loop: asyncio.AbstractEventLoop
-
     @classmethod
-    def on_executor_start(cls):
+    def on_node_executor_start(cls):
         cls.loop = asyncio.new_event_loop()
         cls.loop.run_until_complete(db.connect())
         cls.agent_server_client = get_agent_server_client()
 
     @classmethod
-    def on_start_execution(cls, q: ExecutionQueue, data: Execution) -> bool:
+    def on_node_execution(cls, q: ExecutionQueue[NodeExecution], data: NodeExecution):
         prefix = get_log_prefix(data.graph_exec_id, data.node_exec_id)
         try:
-            logger.warning(f"{prefix} Start execution")
+            logger.warning(f"{prefix} Start node execution")
             for execution in execute_node(cls.loop, cls.agent_server_client, data):
                 q.add(execution)
-            return True
+            logger.warning(f"{prefix} Finished node execution")
+        except Exception as e:
+            logger.exception(f"{prefix} Failed node execution: {e}")
+
+    @classmethod
+    def on_graph_executor_start(cls):
+        cls.pool_size = Config().num_node_workers
+        cls.executor = ProcessPoolExecutor(
+            max_workers=cls.pool_size,
+            initializer=cls.on_node_executor_start,
+        )
+        logger.warning(f"Graph executor started with max-{cls.pool_size} node workers.")
+
+    @classmethod
+    def on_graph_execution(cls, graph_data: GraphExecution):
```

Comment:
Would it be possible to add some basic doc strings on the functions to help people understand how the executors / manager work please

Reply:
Added